modules/packages/EpochManager.chpl

```diff
@@ -0,0 +1,1147 @@
+    pragma "no doc"
+    var free_list : unmanaged LockFreeQueue(unmanaged _token);
+
+    //  Collection of objects marked deleted
+    pragma "no doc"
+    var limbo_list : [1..EBR_EPOCHS] unmanaged LimboList();
+
+    /*
+      Default initialize the manager.
+    */
+    proc init() {
+      allocated_list = new unmanaged LockFreeLinkedList(unmanaged _token);
+      free_list = new unmanaged LockFreeQueue(unmanaged _token, false);
+      this.complete();
+
+      // Initialise the free list pool with here.maxTaskPar tokens
+      forall i in 0..#here.maxTaskPar {
+        var tok = new unmanaged _token();
+        allocated_list.append(tok);
+        free_list.enqueue(tok);
+      }
+      global_epoch.write(1);
+      forall i in 1..EBR_EPOCHS do
+        limbo_list[i] = new unmanaged LimboList();
+    }
+
+    /*
+      Register a task.
+
+      :returns: A handle to the manager
+    */
+    proc register() : owned TokenWrapper { // Should be called only once
+      var tok = free_list.dequeue();
+      if (tok == nil) {
+        tok = new unmanaged _token();
+        allocated_list.append(tok);
+      }
+      tok.is_registered.write(true);
+      // return tok;
+      return new owned TokenWrapper(tok, this:unmanaged);
+    }
+
+    pragma "no doc"
+    proc unregister(tok: unmanaged _token) {
+      if (tok.is_registered.read()) {
+        unpin(tok);
+        free_list.enqueue(tok);
+        tok.is_registered.write(false);
+      }
+    }
+
+    pragma "no doc"
+    proc pin(tok: unmanaged _token) {
+      // An inactive task has local_epoch set to 0. A value other than 0
+      // implies active task
+      if (tok.local_epoch.read() == INACTIVE) then
+        tok.local_epoch.write(global_epoch.read());
+    }
+
+    pragma "no doc"
+    proc unpin(tok: unmanaged _token) {
+      tok.local_epoch.write(INACTIVE);
+    }
+
+    // Attempt to announce a new epoch
+    pragma "no doc"
+    proc try_advance() : uint {
+      var epoch = global_epoch.read();
+      for tok in allocated_list {
+        var local_epoch = tok.local_epoch.read();
+        if (local_epoch > 0 && local_epoch != epoch) then
+          return 0;
+      }
+
+      // Advance the global epoch
+      epoch = (epoch % EBR_EPOCHS) + 1;
+      global_epoch.write(epoch);
+
+      // Return epoch which is safe to be reclaimed. It is safe to
+      // reclaim from e-2 epoch
+      select epoch {
+        when 1 do return EBR_EPOCHS - 1;
+        when 2 do return EBR_EPOCHS;
+        otherwise do return epoch - 2;
+      }
+    }
+
+    pragma "no doc"
+    proc delete_obj(tok : unmanaged _token, x : unmanaged object) {
+      var del_epoch = tok.local_epoch.read();
+      if (del_epoch == 0) {
+        writeln("Bad local epoch! Please pin! Using global epoch!");
+        del_epoch = global_epoch.read();
+      }
+      limbo_list[del_epoch].push(x);
+    }
+
+    /*
+      Try to announce a new epoch. If successful, reclaim objects which are
+      safe to reclaim
```

Comment:
Is it legal to call this function when pinned? It seems the documentation should say something about that.

Reply:
It is legal to call it in or outside of a read-side critical section (while inside a `pin`).

Reply:
Let's say that explicitly in the docs.